src/main/java/com/avispl/symphony/dal/infrastructure/management/biamp/workplace/BiampWorkplaceCommunicator.java

```diff
@@ -0,0 +1,862 @@
+				}
+				retrieveToken();
+			}
+		}
+
+		/**
+		 * Retrieves a new access token using the provided or available refresh token.
+		 *
+		 * @throws FailedLoginException if the token retrieval fails due to invalid credentials or other errors.
+		 */
+		private void retrieveToken() throws FailedLoginException {
+				try{
+					String token = Optional.ofNullable(objToken.getRefreshToken()).orElse(refreshToken);
+
+					if (StringUtils.isNullOrEmpty(token)) {
+						throw new RuntimeException("No valid refresh token available. Login might be required.");
+					}
+					MultiValueMap<String, String> requestBody = new LinkedMultiValueMap<>();
+					requestBody.put(BiampWorkplaceConstant.GRANT_TYPE, Collections.singletonList(BiampWorkplaceConstant.REFRESH_TOKEN));
+					requestBody.put(BiampWorkplaceConstant.CLIENT_ID, Collections.singletonList(this.getLogin()));
+					requestBody.put(BiampWorkplaceConstant.REFRESH_TOKEN, Collections.singletonList(token));
+					JsonNode result = doPost(BiampWorkplaceCommand.SIMULATOR_GET_TOKEN, requestBody, JsonNode.class);
+
+					objToken = objectMapper.treeToValue(result.get("response"), RefreshToken.class);
+					this.loginInfo.setToken(objToken.getAccessToken());
+					this.accessToken = objToken.getAccessToken();
+					this.loginInfo.setLoginDateTime(System.currentTimeMillis());
+
+				} catch (Exception e) {
+					throw new FailedLoginException("Invalid token");
+				}
+		}
+
+		/**
+		 * Create a button.
+		 *
+		 * @param name name of the button
+		 * @param label label of the button
+		 * @param labelPressed label of the button after pressing it
+		 * @param gracePeriod grace period of button
+		 * @return This returns the instance of {@link AdvancedControllableProperty} type Button.
+		 */
+		private AdvancedControllableProperty createButton(String name, String label, String labelPressed, long gracePeriod) {
+			AdvancedControllableProperty.Button button = new AdvancedControllableProperty.Button();
+			button.setLabel(label);
+			button.setLabelPressed(labelPressed);
+			button.setGracePeriod(gracePeriod);
+			return new AdvancedControllableProperty(name, new Date(), button, BiampWorkplaceConstant.EMPTY);
+		}
+
+		/**
+		 * Add addAdvancedControlProperties if advancedControllableProperties different empty
+		 *
+		 * @param advancedControllableProperties advancedControllableProperties is the list that store all controllable properties
+		 * @param stats store all statistics
+		 * @param property the property is item advancedControllableProperties
+		 * @throws IllegalStateException when exception occur
+		 */
+		private void addAdvancedControlProperties(List<AdvancedControllableProperty> advancedControllableProperties, Map<String, String> stats, AdvancedControllableProperty property, String value) {
+			if (property != null) {
+				advancedControllableProperties.removeIf(controllableProperty -> controllableProperty.getName().equals(property.getName()));
```

Comment:
is there a need to remove it every time if it exists? maybe in this case it'd be ok just to update the existing control's timestamp?